src/main/java/org/kitteh/irc/client/library/implementation/STSHandler.java

```diff
@@ -0,0 +1,220 @@
+/*
+ * * Copyright (C) 2013-2016 Matt Baxter http://kitteh.org
+ *
+ * Permission is hereby granted, free of charge, to any person
+ * obtaining a copy of this software and associated documentation
+ * files (the "Software"), to deal in the Software without
+ * restriction, including without limitation the rights to use, copy,
+ * modify, merge, publish, distribute, sublicense, and/or sell copies
+ * of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package org.kitteh.irc.client.library.implementation;
+
+import net.engio.mbassy.listener.Handler;
+import org.kitteh.irc.client.library.element.CapabilityState;
+import org.kitteh.irc.client.library.element.ServerMessage;
+import org.kitteh.irc.client.library.event.capabilities.CapabilitiesNewSupportedEvent;
+import org.kitteh.irc.client.library.event.capabilities.CapabilitiesSupportedListEvent;
+import org.kitteh.irc.client.library.event.client.ClientConnectionClosedEvent;
+import org.kitteh.irc.client.library.exception.KittehServerMessageException;
+import org.kitteh.irc.client.library.feature.sts.STSClientState;
+import org.kitteh.irc.client.library.feature.sts.STSMachine;
+import org.kitteh.irc.client.library.feature.sts.STSPolicy;
+import org.kitteh.irc.client.library.feature.sts.STSStorageManager;
+import org.kitteh.irc.client.library.util.STSUtil;
+
+import java.net.InetSocketAddress;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+/**
+ * Class for handling the STS capability,
+ * returned in the CAP LS 302 response.
+ */
+class STSHandler {
+    private final STSMachine machine;
+    private final InternalClient client;
+    private boolean isSecure;
+    private static final Predicate<CapabilityState> STS_CAPABILITY_PREDICATE = c -> c.getName().equals("sts");
+
+    /**
+     * Creates the event handler for STS.
+     *
+     * @param machine the STS FSM
+     * @param client the IRC client
+     */
+    STSHandler(STSMachine machine, InternalClient client) {
+        this.machine = machine;
+        this.client = client;
+        this.isSecure = client.getConfig().getNotNull(Config.SSL);
```

Comment:
Perhaps extract `client.getConfig().getNotNull(Config.SSL)` to a method on `InternalClient` since it seems to be used quite a bit.